### PR TITLE
refactor: migrate scratch-blocks to Typescript

### PR DESCRIPTION
**Pull request by gonfunko**
This PR migrates the scratch-blocks codebase to Typescript, and cleans up/modernizes a few files. Two files are unconverted:

* checkable_continuous_flyout.js, because it calls `layout_()` on VerticalFlyout (bypassing its parent class), which is a protected method. Although it is a subclass, because its parent class (ContinuousFlyout) is JS, TS doesn't know this and complains. This is a temporary workaround that can be removed once the Continuous Toolbox plugin is updated to account for the new flyout API, at which point this file can be converted.
* scratch_comment_bubble.js, because `ISelectable` requires a public `workspace` field. `CommentView`, the superclass of `ScratchCommentBubble`, has a `workspace` field, but it's private, so `CommentView` can't make it public to implement the interface. This will either need to be made public in core, or have the field in core named something else.

### CODE REVIEW

File: src/recyclable_block_flyout_inflater.ts [hunk 1]
```diff
@@ -13,26 +13,27 @@ import { CheckboxBubble } from "./checkbox_bubble";
...
+   * @returns The newly created block.
    */
-  load(state, flyoutWorkspace) {
+  load(
+    state: Blockly.utils.toolbox.BlockInfo,
+    flyoutWorkspace: Blockly.WorkspaceSvg
+  ): Blockly.IBoundedElement {
     const block = super.load(state, flyoutWorkspace);
-    if (block.checkboxInFlyout) {
+    if ("checkboxInFlyout" in block && block.checkboxInFlyout) {
```
Comment (rachel-fenichel) [inline src/recyclable_block_flyout_inflater.ts:36]: Why the behaviour change here?

File: src/fields/scratch_field_number.ts [hunk 2]
```diff
@@ -45,7 +45,10 @@ import { Colours } from "../colours";
  * @extends {Blockly.FieldTextInput}
  * @constructor
  */
-class FieldNumberPicker extends Blockly.FieldTextInput {
+class ScratchFieldNumber extends Blockly.FieldTextInput {
+  private negativeAllowed_ = true;
```
Comment (rachel-fenichel) [inline src/fields/scratch_field_number.ts:49]: Nit for later: if private, you can take away the underscores.

Comment (rachel-fenichel) [reply src/fields/scratch_field_number.ts:49]: Same for some methods in this file.

**Review by rachel-fenichel — COMMENTED** [review]
LGTM but please ask CWF to look at the structure of it and sign off.

For later cleanup: many fields have a lot of underscores on private properties, which can now be renamed.

File: src/recyclable_block_flyout_inflater.ts (hunk 1 above)
Comment (gonfunko) [reply src/recyclable_block_flyout_inflater.ts:36]: `checkboxInFlyout` isn't on Block/BlockSvg, it's added by an extension, so we have to check for its presence to satisfy TS.

File: src/fields/scratch_field_number.ts (hunk 2 above)
Comment (gonfunko) [reply src/fields/scratch_field_number.ts:49]: Filed #225 to track.

**Comment by gonfunko** [comment]
@cwillisf let me know if this all looks good to you!

File: src/blocks/data.ts [hunk 3]
```diff
@@ -518,51 +481,57 @@ Blockly.Blocks["data_hidelist"] = {
...
+   * @param options List of menu options to add to.
    */
-  customContextMenu: function (options) {
-    var fieldName = "VARIABLE";
+  customContextMenu: function (
+    this: Blockly.Block,
+    options: Array<
+      | Blockly.ContextMenuRegistry.ContextMenuOption
+      | Blockly.ContextMenuRegistry.LegacyContextMenuOption
+    >
```
Comment (cwillisf) [inline src/blocks/data.ts:496]: :muscle:

File: src/scratch_variable_model.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+ * Class that represents a variable with extra fields for Scratch.
+ */
+export class ScratchVariableModel extends Blockly.VariableModel {
+  constructor(
+    workspace: Blockly.Workspace,
+    name: string,
+    type: string,
+    id: string,
+    public isLocal = false,
+    public isCloud = false
```
Comment (cwillisf) [inline src/scratch_variable_model.ts:19]: :exploding_head:
I didn't know about this feature... wild!

**Review by cwillisf — APPROVED** [review]
Exciting!

**Pull request merged**